[DOCS] Create New Expectation Classes Conceptual Guide (#8004)
great-expectations/great_expectations: docs/docusaurus/docs/conceptual_guides/expectation_classes.md

---
title: Expectation classes
---

> In our daily lives and our data, we expect different things from different types of objects. After all, it would be very alarming and disturbing if a chair suddenly behaved like a cat! It is therefore entirely reasonable to expect different results when we evaluate the contents of an entire table and when we evaluate the contents of a single column. Expectation classes have been created to help make sure Custom Expectations return the expected results. This document provides an overview of the available Expectation classes, why they are helpful, and when they should be used.

Reviewer: ```suggestion
In our daily lives and our data, we expect different things from different types of objects. After all, it would be both alarming and disturbing if a chair suddenly behaved like a cat! Similarly, it is entirely reasonable to expect different results when we evaluate the contents of an entire table and when we evaluate the contents of a single column. Expectation classes have been created to help make sure Custom Expectations return the expected results. This document provides an overview of the available Expectation classes, why they are helpful, and when they should be used.
```

## Class hierarchy

The following is the Expectation class hierarchy:

```text
Expectation
    BatchExpectation
        ColumnAggregateExpectation
        ColumnMapExpectation
            RegexBasedColumnMapExpectation
            SetBasedColumnMapExpectation
        ColumnPairMapExpectation
        MulticolumnMapExpectation
        QueryExpectation
```

Most Expectations are a combination of a Domain (Batch, Column, ColumnPair, Multicolumns) and an approach (Map or Aggregate). In some cases, the Expectation classes include a prefix such as `RegexBasedColumnMapExpectation`. There are also two classes that don’t follow the standard naming convention; `BatchExpectations` and `QueryExpectations`.

## Expectation Domain types

> Domains provide a way to address a specific set of data, such as a batch within a table, or a column. Domains do this by describing the data locale. The data locale is the conceptual equivalent of “data that arrived last Tuesday in the UserEvents table in the Redshift database,” or “the timestamp column in the User's table in the Redshift database”.

Reviewer: ```suggestion
Domains provide a way to address a specific set of data, such as a Batch within a table, or a column within a Batch. Domains do this by describing the data locale. The data locale is the conceptual equivalent of “data that arrived last Tuesday in the UserEvents table in the Redshift database,” or “the timestamp column in the User's table in the Redshift database”.
```

The following are the four Expectation Domains:

- Batch
- Column
- ColumnPair
- Multicolumn

### ColumnPair

A ColumnPair is the special case of a MultiColumn where the number of columns equals two. A Column Expectation is the special case where the number equals one.

> From a software engineering perspective, however, there are meaningful differences between Expectations with different domains. Specifically, all Column Expectations accept a `column` argument, all ColumnPair Expectations accept a pair of arguments, usually named `column_A` and `column_B`, and all MultiColumn Expectations accept a `column_list` argument.

Author: ```suggestion
From a software engineering perspective, there are meaningful differences between Expectations with different domains. Specifically, all Column Expectations accept a `column` argument, all ColumnPair Expectations accept a pair of arguments, usually named `column_A` and `column_B`, and all MultiColumn Expectations accept a `column_list` argument.
```

Aggregate Expectations calculate summary statistics across Batches of data. As a result, they can be a computationally efficient way to gain insight into the overall behavior of a dataset and can provide a useful foundation for identifying trends, patterns, and outliers. However, because Aggregate Expectations do not verify individual rows of data, they can't identify specific data issues.

## Map Expectations

Map Expectations are evaluated on a row-by-row basis and each row is checked independently. For example, [`expect_column_values_to_not_be_null`](https://greatexpectations.io/expectations/expect_column_values_to_not_be_null?filterType=Backend%20support&gotoPage=1&showFilters=true&viewType=Summary), [`expect_column_values_to_be_in_set`](https://greatexpectations.io/expectations/expect_column_values_to_not_be_null?filterType=Backend%20support&gotoPage=1&showFilters=true&viewType=Summary), [` expect_column_pair_values_to_be_equal`](https://greatexpectations.io/expectations/expect_column_pair_values_to_be_equal?filterType=Backend%20support&gotoPage=1&showFilters=true&viewType=Summary).

![Map Expectations](../images/map_expectations.png)

> Map Expectations are useful when you want to be certain that the content of a given dataset is correct. If you’re validating data within a pipeline, Map Expectations can help you identify invalid rows, remove invalid rows from the dataset, and process the remaining data. Unfortunately, Map Expectations evaluate every row of data and this can be computationally intensive.

Reviewer: ```suggestion
Map Expectations are useful when you want to be certain that the content of a given dataset is correct. If you’re validating data within a pipeline, Map Expectations can help you identify invalid rows, remove invalid rows from the dataset, and process the remaining data. Unfortunately, because Map Expectations evaluate every row of data they can be computationally intensive.
```

> Map Expectations are useful when you want to be certain that the content of a given dataset is correct. If you’re validating data within a pipeline, Map Expectations can help you identify invalid rows, remove invalid rows from the dataset, and process the remaining data. Unfortunately, because Map Expectations evaluate every row of data they can be computationally intensive.

Author: ```suggestion
Map Expectations are useful when you want to be certain that the content of a given dataset is correct. If you’re validating data within a pipeline, Map Expectations can help you identify invalid rows, remove invalid rows from the dataset, and process the remaining data. Unfortunately, because Map Expectations evaluate every row of data, they can be computationally intensive.
```

## Subclasses

Beyond the `[Domain][Approach]Expectation` naming convention, the specialized subclasses `RegexBasedColumnMapExpectation` and `SetBasedColumnMapExpectation` are supported. These extend `ColumnMapExpectation` and make it easier to define Expectations based on regexes and sets. For more information, see [How to create a Custom Regex-Based Column Map Expectation](../guides/expectations/creating_custom_expectations/how_to_create_custom_regex_based_column_map_expectations.md) and [How to create a Custom Set-Based Column Map Expectation](../guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md).

## BatchExpectations

BatchExpectations do not currently have a special subclass for Map Expectations. Essentially, BatchMapExpectations would apply row-by-row validation to all the columns in a Batch. When there is demand for this Expectation type, the class hierarchy will be refactored to accommodate it.

## QueryExpectations

> QueryExpectations allow you to set Expectations against the results of custom SQL or Spark queries. Query Expectations can be useful if you’re comfortable working in SQL or Spark or a specific dialect. They can also allow you to embed arbitrarily complex logic in your Expectations, such as combining data from multiple tables, or applying complex logic within a query.

Author: ```suggestion
QueryExpectations allow you to set Expectations against the results of custom SQL or Spark queries. QueryExpectations can be useful if you’re comfortable working in SQL or Spark or a specific dialect. They can also allow you to embed arbitrarily complex logic in your Expectations, such as combining data from multiple tables, or applying complex logic within a query.
```
